Source/WebCore/rendering/updating/RenderTreeUpdater.cpp

```diff
@@ -570,12 +570,37 @@ bool RenderTreeUpdater::textRendererIsNeeded(const Text& textNode)
     if (is<RenderText>(renderingParent.previousChildRenderer))
         return true;
     // This text node has nothing but white space. We may still need a renderer in some cases.
-    if (parentRenderer.isRenderTable() || parentRenderer.isRenderTableRow() || parentRenderer.isRenderTableSection() || parentRenderer.isRenderTableCol() || parentRenderer.isRenderFrameSet() || parentRenderer.isRenderGrid() || (parentRenderer.isRenderFlexibleBox() && !parentRenderer.isRenderButton()))
+    auto* previousRenderer = renderingParent.previousChildRenderer;
+    if (parentRenderer.isRenderTableRow() || parentRenderer.isRenderTableSection() || parentRenderer.isRenderTableCol() || parentRenderer.isRenderFrameSet() || parentRenderer.isRenderGrid() || (parentRenderer.isRenderFlexibleBox() && !parentRenderer.isRenderButton()))
         return false;
+    if (parentRenderer.isRenderTable()) {
+        if (!previousRenderer)
+            return false;
+
+        // Suppress whitespace after table-internal or block-level elements.
+        if (previousRenderer->isRenderTableRow() || previousRenderer->isRenderTableSection() || previousRenderer->isRenderTableCol() || previousRenderer->isRenderTableCaption())
+            return false;
+
+        // Check if the element is a form control (input, button, etc.).
+        if (is<RenderElement>(*previousRenderer)) {
+            auto& renderElement = downcast<RenderElement>(*previousRenderer);
+            if (auto* element = renderElement.element()) {
+                if (element->hasTagName(HTMLNames::inputTag) || element->hasTagName(HTMLNames::buttonTag))
+                    return false;
+            }
+        }
+
+        // Preserve whitespace only for inline-block and inline-table.
+        auto display = previousRenderer->style().display();
+        if (display == DisplayType::InlineBlock || display == DisplayType::InlineTable)
```

Comment:
Just add `|| display == DisplayType::Inline` and it will fix failure where it is just `<span>` (inline layout).

Reply:
but it will fail - http://wpt.live/css/css-display/display-contents-dynamic-table-002-inline.html